libsolidity/analysis/IntegerOverflowBounds.cpp

```diff
@@ -0,0 +1,267 @@
+	solAssert(false, "IntegerOverflowBounds fits only IntegerType");
+}
+
+bool IntegerOverflowBounds::fitsType(IntegerType const& type)
+{
+	bool isSigned = type.isSigned();
+	uint16_t bytes = type.storageBytes();
+	uint32_t exponent = bytes * (isSigned ? ((bytes * 8) - 1) : (bytes * 8));
+	dev::bigint allowedMin;
+	dev::bigint allowedMax;
+	if (isSigned)
+	{
+		allowedMin = 1;
+		allowedMin = allowedMin << exponent;
+		allowedMin = -allowedMin;
+
+		allowedMax = 1;
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
+	}
+	else
+	{
+		allowedMin = 0;
+		allowedMax = 1;
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
+	}
+	return (min >= allowedMin && max <= allowedMax);
+}
+
+void IntegerOverflowBounds::setToMaximumRange(Type const& type)
+{
+	if (type.category() == Type::Category::Integer)
+		setToMaximumRange(static_cast<IntegerType const&>(type));
+	else if (type.category() == Type::Category::RationalNumber)
+		setToMaximumRange(static_cast<RationalNumberType const&>(type));
+	else
+		solAssert(false, string("IntegerOverflowBounds doesn't support type ")+type.toString());
+}
+
+void IntegerOverflowBounds::setToMaximumRange(IntegerType const& type)
+{
+	bool isSigned = type.isSigned();
+	uint16_t bytes = type.storageBytes();
+	uint32_t exponent = bytes * (isSigned ? ((bytes * 8) - 1) : (bytes * 8));
+	dev::bigint allowedMin;
+	dev::bigint allowedMax;
+	if (isSigned)
+	{
+		allowedMin = 1;
+		allowedMin = allowedMin << exponent;
+		allowedMin = -allowedMin;
+
+		allowedMax = 1;
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
+	}
+	else
+	{
+		allowedMin = 0;
+		allowedMax = 1;
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
+	}
+	min = allowedMin;
+	max = allowedMax;
+}
+
+void IntegerOverflowBounds::setToMaximumRange(RationalNumberType const& type)
+{
+	min = numerator(type.value());
+	max = numerator(type.value());
+}
+
+void IntegerOverflowBounds::setToDefaultValue()
+{
+	min = 0;
+	max = 0;
+}
+
+void IntegerOverflowBounds::constrain(OverflowBounds const& bounds)
+{
+	auto const integerBounds = dynamic_cast<IntegerOverflowBounds const*>(&bounds);
+	solAssert(integerBounds, "Only accepting assignment from Integer and Rational");
+
+	//A constraint is an enforced and strict bounds produced from a logical assertion.
+	//Thus we will retract the range until it fits the described bounds.
+	if (min < integerBounds->min)
+		min = integerBounds->min;
+	if (max > integerBounds->max)
+		max = integerBounds->max;
+}
+
+OverflowBounds& IntegerOverflowBounds::operator=(OverflowBounds const& bounds)
+{
+	auto const integerBounds = dynamic_cast<IntegerOverflowBounds const*>(&bounds);
+	solAssert(integerBounds, "Only accepting assignment from Integer and Rational");
+	operator=(*integerBounds);
+	return *this;
+}
+
+OverflowBounds& IntegerOverflowBounds::operator=(RationalNumberType const& number)
+{
+	min = numerator(number.value());
```

Comment:
Please handle the case (assertion could be fine for now) where `denominator(number) != 1`